File: clojure/src/step9_interop.clj
(ns step9-interop
    (:refer-clojure :exclude [macroexpand])
    (:require [clojure.repl]
              [types]
              [readline]
              [reader]))

(declare EVAL)

;; read
(defn READ [& [strng]]
  (let [line (if strng strng (read-line))]
    (reader/read-string strng)))

;; eval
(defn is-pair [x]
  (and (sequential? x) (> (count x) 0)))

(defn quasiquote [ast]
  (cond
    (not (is-pair ast))
    (list 'quote ast)

    (= 'unquote (first ast))
    (second ast)

    (and (is-pair (first ast)) (= 'splice-unquote (ffirst ast)))
    (list 'concat (-> ast first second) (quasiquote (rest ast)))

    :else
    (list 'cons (quasiquote (first ast)) (quasiquote (rest ast)))))

(defn is-macro-call [ast env]
  (and (seq? ast)
       (symbol? (first ast))
       (types/env-find env (first ast))
       (:ismacro (meta (types/env-get env (first ast))))))

(defn macroexpand [ast env]
  (loop [ast ast]
    (if (is-macro-call ast env)
      (let [mac (types/env-get env (first ast))]
        (recur (apply mac (rest ast))))
      ast)))

(defn eval-ast [ast env]
  (cond
    (symbol? ast) (types/env-get env ast)
   
    (seq? ast)    (doall (map #(EVAL % env) ast))

    (vector? ast) (vec (doall (map #(EVAL % env) ast)))
   
    (map? ast)    (apply hash-map (doall (map #(EVAL % env)
                                              (mapcat identity ast))))

    :else         ast))

(defn EVAL [ast env]
  (loop [ast ast
         env env]
    ;;(prn "EVAL" ast (keys @env)) (flush)
    (if (not (seq? ast))
      (eval-ast ast env)
  
      ;; apply list
      (let [ast (macroexpand ast env)]
        (if (not (seq? ast))
          ast

          (let [[a0 a1 a2 a3] ast]
            (condp = a0
              'def!
              (types/env-set env a1 (EVAL a2 env))
      
              'let*
              (let [let-env (types/env env)]
                (doseq [[b e] (partition 2 a1)]
                  (types/env-set let-env b (EVAL e let-env)))
                (EVAL a2 let-env))
    
              'quote
              a1
    
              'quasiquote
              (EVAL (quasiquote a1) env)
    
              'defmacro!
              (let [func (with-meta (EVAL a2 env)
                                    {:ismacro true})]
                (types/env-set env a1 func))
    
              'macroexpand
              (macroexpand a1 env)
      
              'clj*
              (eval (reader/read-string a1))
      
              'do
              (do (eval-ast (->> ast (drop-last) (drop 1)) env)
                  (recur (last ast) env))
      
              'if
              (let [cond (EVAL a1 env)]
                (if (or (= cond nil) (= cond false))
                  (if (> (count ast) 2)
                    (recur a3 env)
                    nil)
                  (recur a2 env)))
      
              'fn*
              ^{:expression a2
                :environment env
                :parameters a1}
              (fn [& args]
                (EVAL a2 (types/env env a1 args)))
      
              ;; apply
              (let [el (eval-ast ast env)
                       f (first el)
                       args (rest el)
                       {:keys [expression environment parameters]} (meta f)]
                (if expression
                  (recur expression (types/env environment parameters args))
                  (apply f args))))))))))

;; print
(defn PRINT [exp] (pr-str exp))

;; repl
(def repl-env (types/env))
(defn rep
  [strng]
  (PRINT (EVAL (READ strng) repl-env)))

(defn _ref [k,v] (types/env-set repl-env k v))

;; Import types related functions
(doseq [[k v] types/types_ns] (_ref k v))

;; Defined using the language itself
(_ref 'read-string reader/read-string)
(_ref 'eval (fn [ast] (EVAL ast repl-env)))
(_ref 'slurp slurp)
(_ref 'slurp-do (fn [f] (str "(do " (slurp f) ")")))

(rep "(def! not (fn* [a] (if a false true)))")
(rep "(def! load-file (fn* [f] (eval (read-string (slurp-do f)))))")

(defn -main [& args]
  (if args
    (rep (str "(load-file \"" (first args) "\")"))
    (loop []
      (let [line (readline/readline "user> ")]
        (when line
          (when-not (re-seq #"^\s*$|^\s*;.*$" line) ; blank/comment
            (try
              (println (rep line))
              (catch Throwable e
                (clojure.repl/pst e))))
          (recur))))))

File: php/step9_interop.php
<?php

require_once 'readline.php';
require_once 'types.php';
require_once 'reader.php';

// read
function READ($str) {
    return read_str($str);
}

// eval
function is_pair($x) {
    return sequential_Q($x) and count($x) > 0;
}

function quasiquote($ast) {
    if (!is_pair($ast)) {
        return new_list(new_symbol("quote"), $ast);
    } elseif (symbol_Q($ast[0]) && $ast[0]->value === 'unquote') {
        return $ast[1];
    } elseif (is_pair($ast[0]) && symbol_Q($ast[0][0]) &&
              $ast[0][0]->value === 'splice-unquote') {
        return new_list(new_symbol("concat"), $ast[0][1],
                        quasiquote($ast->slice(1)));
    } else {
        return new_list(new_symbol("cons"), quasiquote($ast[0]),
                        quasiquote($ast->slice(1)));
    }
}

function is_macro_call($ast, $env) {
    return is_pair($ast) &&
           symbol_Q($ast[0]) &&
           $env->find($ast[0]->value) &&
           $env->get($ast[0]->value)->ismacro;
}

function macroexpand($ast, $env) {
    while (is_macro_call($ast, $env)) {
        $mac = $env->get($ast[0]->value);
        $args = array_slice($ast->getArrayCopy(),1);
        $ast = $mac->apply($args);
    }
    return $ast;
}

function eval_ast($ast, $env) {
    if (symbol_Q($ast)) {
        return $env->get($ast->value);
    } elseif (list_Q($ast) || vector_Q($ast)) {
        if (list_Q($ast)) {
            $el = new_list();
        } else {
            $el = new_vector();
        }
        foreach ($ast as $a) { $el[] = MAL_EVAL($a, $env); }
        return $el;
    } elseif (hash_map_Q($ast)) {
        $new_hm = new_hash_map();
        foreach (array_keys($ast->getArrayCopy()) as $key) {
            $new_hm[$key] = MAL_EVAL($ast[$key], $env);
        }
        return $new_hm;
    } else {
        return $ast;
    }
}

function MAL_EVAL($ast, $env) {
    while (true) {
        #echo "MAL_EVAL: " . _pr_str($ast) . "\n";
        if (!list_Q($ast)) {
            return eval_ast($ast, $env);
        }

        // apply list
        $ast = macroexpand($ast, $env);
        if (!list_Q($ast)) { return $ast; }

        $a0 = $ast[0];
        $a0v = (symbol_Q($a0) ? $a0->value : $a0);
        switch ($a0v) {
        case "def!":
            $res = MAL_EVAL($ast[2], $env);
            return $env->set($ast[1]->value, $res);
        case "let*":
            $a1 = $ast[1];
            $let_env = new Env($env);
            for ($i=0; $i < count($a1); $i+=2) {
                $let_env->set($a1[$i]->value, MAL_EVAL($a1[$i+1], $let_env));
            }
            return MAL_EVAL($ast[2], $let_env);
        case "quote":
            return $ast[1];
        case "quasiquote":
            return MAL_EVAL(quasiquote($ast[1]), $env);
        case "defmacro!":
            $func = MAL_EVAL($ast[2], $env);
            $func->ismacro = true;
            return $env->set($ast[1]->value, $func);
        case "macroexpand":
            return macroexpand($ast[1], $env);
        case "php*":
            return eval($ast[1]);
        case "do":
            eval_ast($ast->slice(1, -1), $env);
            $ast = $ast[count($ast)-1];
            break;
        case "if":
            $cond = MAL_EVAL($ast[1], $env);
            if ($cond === NULL || $cond === false) {
                if (count($ast) === 4) { $ast = $ast[3]; }
                else                   { $ast = NULL; }
            } else {
                $ast = $ast[2];
            }
            break;
        case "fn*":
            return new_function('MAL_EVAL', 'native',
                                new_hash_map('exp', $ast[2],
                                             'env', $env,
                                             'params', $ast[1]));
        default:
            $el = eval_ast($ast, $env);
            $f = $el[0];
            $args = array_slice($el->getArrayCopy(), 1);
            if ($f->type === 'native') {
                $ast = $f->meta['exp'];
                $env = new Env($f->meta['env'], $f->meta['params'], $args);
            } else {
                return $f->apply($args);
            }
        }
    }
}

// print
function MAL_PRINT($exp) {
    return _pr_str($exp, True) . "\n";
}

// repl
$repl_env = new Env(NULL);
function rep($str) {
    global $repl_env;
    return MAL_PRINT(MAL_EVAL(READ($str), $repl_env));
}
function _ref($k, $v) {
    global $repl_env;
    $repl_env->set($k, new_function($v));
}
// Import types functions
foreach ($types_ns as $k=>$v) { _ref($k, $v); }

_ref('read-string', 'read_str');
_ref('eval', function($ast) {
    global $repl_env; return MAL_EVAL($ast, $repl_env);
});
_ref('slurp', function($f) {
    return file_get_contents($f);
});

// Defined using the language itself
rep("(def! not (fn* (a) (if a false true)))");
rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \")\")))))");

if (count($argv) > 1) {
    for ($i=1; $i < count($argv); $i++) {
        rep('(load-file "' . $argv[$i] . '")');
    }
} else {
    do {
        try {
            $line = mal_readline("user> ");
            if ($line === NULL) { break; }
            if ($line !== "") {
                print(rep($line));
            }
        } catch (BlankException $e) {
            continue;
        } catch (Exception $e) {
            echo "Error: " . $e->getMessage() . "\n";
            echo $e->getTraceAsString() . "\n";
        }
    } while (true);
}

?> 
